src/crystal/event_loop/io_uring.cr

```diff
@@ -0,0 +1,483 @@
+      when Errno::EBADF
+        raise IO::Error.new("File not open for reading", target: file_descriptor)
+      else
+        raise IO::Error.from_os_error("read", errno, target: file_descriptor)
+      end
+    end
+  end
+
+  def wait_readable(file_descriptor : System::FileDescriptor) : Nil
+    async_poll(file_descriptor, LibC::POLLIN | LibC::POLLRDHUP, file_descriptor.@read_timeout) { "Read timed out" }
+  end
+
+  def write(file_descriptor : System::FileDescriptor, slice : Bytes) : Int32
+    async_rw(LibC::IORING_OP_WRITE, file_descriptor, slice, file_descriptor.@write_timeout) do |errno|
+      case errno
+      when Errno::ECANCELED
+        raise IO::TimeoutError.new("Write timed out")
+      when Errno::EBADF
+        raise IO::Error.new("File not open for writing", target: file_descriptor)
+      else
+        raise IO::Error.from_os_error("write", errno, target: file_descriptor)
+      end
+    end
+  end
+
+  def wait_writable(file_descriptor : System::FileDescriptor) : Nil
+    async_poll(file_descriptor, LibC::POLLOUT, file_descriptor.@write_timeout) { "Write timed out" }
+  end
+
+  def reopened(file_descriptor : System::FileDescriptor) : Nil
+    # TODO: do we need to cancel pending operations?
+  end
+
+  def close(file_descriptor : System::FileDescriptor) : Nil
+    # sync with `FileDescriptor#file_descriptor_close`: prevent actual close
+    fd = file_descriptor.@volatile_fd.swap(-1, :relaxed)
+    return if fd == -1
+
+    async_close(fd) do |sqe|
+      # one thread closing a fd won't interrupt reads or writes happening in
+      # other threads, for example a blocked read on a fifo will keep blocking,
+      # while close would have finished and closed the fd; we thus explicitly
+      # cancel any pending operations on the fd before we try to close
+      #
+      # FIXME: with threads and multiple rings, we'll need to know which rings
+      # have pending operations for the fd (which op/event for each ring) and
+      # tell the rings to cancel said ops (can't just say to cancel all ops for
+      # fd so we can close in parallel)
+      sqe.value.opcode = LibC::IORING_OP_ASYNC_CANCEL
+      sqe.value.sflags.cancel_flags = LibC::IORING_ASYNC_CANCEL_FD
+      sqe.value.fd = fd
+    end
+  end
+
+  # socket interface, see Crystal::EventLoop::Socket
+
+  def read(socket : ::Socket, slice : Bytes) : Int32
+    async_rw(LibC::IORING_OP_READ, socket, slice, socket.@read_timeout) do |errno|
+      case errno
+      when Errno::ECANCELED
+        raise IO::TimeoutError.new("Read timed out")
+      else
+        raise IO::Error.from_os_error("read", errno, target: socket)
+      end
+    end
+  end
+
+  def wait_readable(socket : ::Socket) : Nil
+    async_poll(socket, LibC::POLLIN | LibC::POLLRDHUP, socket.@read_timeout) { "Read timed out" }
+  end
+
+  def write(socket : ::Socket, slice : Bytes) : Int32
+    async_rw(LibC::IORING_OP_WRITE, socket, slice, socket.@write_timeout) do |errno|
+      case errno
```

Comment:
Can `EAGAIN` not happen?

Reply:
It never happened. I believe `io_uring` doesn't care about `O_NONBLOCK` and does its own in-kernel thing.